### ISSUE TITLE
Downloading many episodes at once result in some disappearing even though they successfully downloaded

### ISSUE THREAD

**Issue by mitchdowney** [podverse/podverse-rn#2041]
This is easily reproducible when using the "download full album" button on the new AlbumScreen.

My guess is there some sort of race-condition that leads to the reference we save for the downloaded file overwriting another once being saved at the same time...I'm not sure, needs more investigation.

I wonder if this explains why some people report a massive amount of storage usage by the app, even though they deleted all their episodes?

**Comment by lovegaoshi** [comment]
I also have a cache/download feature, through its through RN-blob-util and its fetching from a different source (ytm):

for blob-util if the ytm download takes too long the connection will drop, an error will emit, I didnt and still dont handle it so the temp files just pile up. for blob-util it has a particular naming scheme (RNBlobTempFiles{}) so I recently did a fs.list and compare any files not present in my LRU cache and deleted them. 
https://github.com/lovegaoshi/azusa-player-mobile/pull/229/files#diff-26b7ddda9d7333e3aaae55428048e7568eb30c1640c48621bcac376a0ef19c63

**Comment by mitchdowney** [comment]
It turned out this was a bug in our `downloader` file. We use `AsyncStorage` to save references to the downloaded files, but when multiple files finish downloading at the same time, AsyncStorage was overwritten in a way that caused other references to be lost.

To resolve this, I added a promise queue that will resolve one-at-a-time for the finish downloading step, so they shouldn't overwrite each other.

This was released in v4.15.0
